### ISSUE TITLE
Jojo no Kimyou na Bouken - Ougon no Kaze (Japan) Files

### ISSUE THREAD

**Issue by num421337** [IcySon55/Kuriimu#355]
Issue: [plugin request]
Console: PS2
Format extension(s): snd,bin,pzz,hit
Type: [unknown]
Game Name: 

First 8 byte of file(s):
<snd> - [4D 4F 4D 4F 04 00 00 00 40] - [MOMO....@]
<bin> - [00 00 00 00 40 00 00 00 05] - [....@....]
<pzz> - [0D 00 00 00 1B 00 00 80 03] - [.......€.]
<hit> - [20 00 00 00 64 00 00 00] - [ ...d...]

More details:
files were found inside a AFS files found on the disc. other files found were identified to be ADX sound files typical to a ps2 game

Sample files (if possible):
<https://mega.nz/#F!4ZBWlQwA!fzeVvfMlc_ZrrLiWCMb8ew>

**Comment by vgf89** [comment]
Oh hey, I was looking for the exact same thing!

Be warned. Bruno Bucciarati's speech is practically all very Japanese puns.

**Comment by vgf89** [comment]
I'm working on reversing the .pzz files and, later, will be working on plugins as well for the different file types in the future. Reversing work first though. I'll add comments to keep up with my work.

Observations for .pzz files:

Byte 0x00: File type identifier
Following bytes (at 0x04, 0x08, 0x0C, 0x10, etc depending on file type) in the header are sizes for each section of the file, in multiples of 2048 bytes. Header size: 0x800 (2048 bytes)
- type 0x1 header:
    - byte 0x01: size of data
- type 0x3 header:
    - byte 0x04: size of data
    - byte 0x08: size of the script
    - byte 0x0C: size of data
- type 0x4 header: just data (sizes are in 0x04, 0x08, 0x0C, and 0x10), we can ignore these files
- type 0x5 header: 
    - bytes 0x04, 0x08: data section sizes
    - byte 0x0C: size of script
    - bytes 0x10, 0x14: data section sizes
- type 0x0D: just data (section size bytes are 0x04, 0x08, 0x0C, 0x28, 0x2C, might be incomplete)

If I remember correctly, the 0x03 pzz files are cutscenes, but don't quote me on that yet. I'm not sure exactly what 0x5 files are yet, but they have timed scripts as well.

Beginning byte of the script = size of section(s) before script, including the header.

Given type 0x3 for instance, and byte 0x04 == 3, script begins at (1+3)*(0x800).
The size of the section allocated for the script is trivial, just 0x800*(value of 0x08). With this knowledge, we can also change the size of the script section if need be by modifying that bit and making sure that section takes up that much space.

**Comment by vgf89** [comment]
@num421337 What .hit files actually had japanese in them? What encoding are you using? Everything else seems to use Shift-JIS, and I'm seeing nothing like that in any of the .hit files.

**Comment by num421337** [comment]
sorry for taking so long to respond. I found text that actually did stuff in the .pzz files. mostly shift-jis stuff. I have the people from Jojo's Colored Adventure in case I need anything translated. the plan was to just link the text to the already finished manga translation. Also I found out that some numbers effect the text boxes. 
![image](https://user-images.githubusercontent.com/29212958/34913865-778d8cc0-f8d5-11e7-8290-dfd66892b930.png)

There's a diagram of what I currently know. Haven't worked on this for a while. 

When I made this thread I had 0 knowledge of what any of the files were so I just was asking if anything had any text.

Basically at this point all in game cutscene type text is changeable. I'm hoping at some point the menu and various other text strings will be modified as well. if not oh well

**Comment by num421337** [comment]
I have been able to edit the text succesfully only with a hex editor. a normal text editor messes with the encoding I assume. I have screenshots of what each numbered bubble looks like 
![image](https://user-images.githubusercontent.com/29212958/34913889-07b33e94-f8d6-11e7-8f82-643ddf3139e0.png) i've also been using afs explorer to be able to insert the modified afs files into the iso without having to entirely rebuild each time.

**Comment by vgf89** [comment]
Seems you figured those parameters out slightly better than I did. IIRC the first and second values are start and stop times (probably in milliseconds). If the second one is zero it should keep the textbox open until the next one shows up, though I need to verify that.

Also IIRC, wxMEdit in text mode works pretty well for doing the actual editing. Making sure the file size stuff stays right can get tedious though... perhaps this coming weekend I'll write a script to pull out and re-insert that chunk of the files to make things easier (that script would also provide reasonable documentation for if/when a plugin gets written). If I get it working I'll upload here.

Speaking of translation, I also have someone who wants to do a from-scratch translation of this game. I might throw the files at her if I get such a script working.

**Comment by num421337** [comment]
The Jojos Colored Adventure Team has told me that most of the text in game is practically identical to their part 5 translation for the manga. Unless formatting is needed to fit the text in the boxes the translation is mostly finished already. I'm hoping we can understand the other codes to possibly add extra boxes or something, or at least eventually find info leading to replacing the menu text

**Comment by onepiecefreak3** [comment]
According to the discussion, it seems we only need to look at the pzz's, so the rest will be ignored.

**Comment by vgf89** [comment]
Here's a Gist of a couple of python script: a pzz script extractor and a script injector.

The extractor looks for .pzz files in the directory pzz_files, checks if the starting byte is 0x03 or 0x05, then write the script sector (minus the 0 padding at the end) to a new file with the name script-filename.pzz.txt.

The injector puts those new script files back into the original .pzz files.

https://gist.github.com/vgf89/d2c83288bb902b51d32dc019f64d94bb

**Comment by num421337** [comment]
This file is essentially the games ELF file but it contains a lot of text that would be a nightmare to edit just using a hex editor. Any possibilities of something to make that a bit easier? (the text generally needs to be the same or similar lengths in order to preserve the order of the other code in it. there is english text in standard format for the credits and SJIS japanese as well
[SLPM_651.zip](https://github.com/IcySon55/Kuriimu/files/1839315/SLPM_651.zip)

**Comment by vgf89** [comment]
I updated the script extractor and injector scripts in the Gist.

0x04 filetype support has been added. The extractor also now makes sure it can find a script in the file before going through with extraction (just a simple regex "[-0-9]+,[-0-9]+,[-0-9]+,[-0-9]+,[-0-9]+,") so we don't end up with junk files. The injector also doesn't fail if it can't find a script to inject, it just skips the file instead.

https://gist.github.com/vgf89/d2c83288bb902b51d32dc019f64d94bb

**Comment by num421337** [comment]
![image](https://user-images.githubusercontent.com/29212958/37888149-b52346e2-3093-11e8-9a49-fe4f76ce9042.png)
![image](https://user-images.githubusercontent.com/29212958/37888155-bed8a8bc-3093-11e8-87e1-308434847a95.png)
![image](https://user-images.githubusercontent.com/29212958/37888167-d31e7ca2-3093-11e8-969f-9fae419fd2a0.png)
Some TIM2 graphics files were found using Game Graphics Studio. But only images like these, no useable ones for translating. It seems likely that other ones are mipmapped as they cannot be read by Rainbow and give a mipmapped image error

**Comment by num421337** [comment]
I was told to look at the debug menu on pcsx2 by someone who has worked with TIM2 files before but I don't know what to look for.
I was also told that it was likely that the files are compressed in some way. no clue how to accomplish this

**Comment by MistaL69** [comment]
Hey, JohnPaulGH, I'm the other dude working on this project. Do you think we could possibly team up together on this? It'd probably speed things up a lot for all of us to collaborate and put what we've done so far together and work from there, plus we can share whatever things we may have learned.

**Comment by JohnPaulGH** [comment]
To MistaL69.
Yes of course! I'm looking for help to finish this as quickly as possible! Look at my blog and tell me what you think, in the comments.
john.jps.arg@gmail.com
https://juanpablosalvatierra.blogspot.com/

**Comment by onepiecefreak3** [comment]
Due to the requested file formats all being very intertwined with the game itself, pzz being a generic contianer for multiple sorts of data, and the availability of scripts that deal with them already, I decide that none of the mentioned formats get introduced into Kuriimu2.
You're free to write your own plugins as soon as your time (or Kuriimu2's framework) allows it. Issue closed.
